packages/zoe/src/zoeService/zoeSeat.js

```diff
@@ -0,0 +1,74 @@
+// @ts-check
+
+import { makePromiseKit } from '@agoric/promise-kit';
+import { makeNotifierKit } from '@agoric/notifier';
+import { assert } from '@agoric/assert';
+import { E } from '@agoric/eventual-send';
+
+import '../types';
+import '../internal-types';
+
+/** @type {MakeZoeSeatAdminKit} */
+export const makeZoeSeatAdminKit = (
+  initialAllocation,
+  instanceAdmin,
+  proposal,
+  brandToPurse,
+  promises = {},
+) => {
+  const payoutPromiseKit = makePromiseKit();
+  const offerResultPromiseKit = promises.offerResult || makePromiseKit();
+  const exitObjPromiseKit = promises.exitObj || makePromiseKit();
+  const { notifier, updater } = makeNotifierKit();
+
+  let currentAllocation = initialAllocation;
+
+  /** @type {ZoeSeatAdmin} */
+  const zoeSeatAdmin = harden({
+    replaceAllocation: replacementAllocation => {
+      assert(
+        instanceAdmin.hasZoeSeatAdmin(zoeSeatAdmin),
+        `Cannot replace allocation. Seat has already exited`,
+      );
+      harden(replacementAllocation);
+      // Merging happens in ZCF, so replacementAllocation can
+      // replace the old allocation entirely.
+      updater.updateState(replacementAllocation);
+      currentAllocation = replacementAllocation;
+    },
+    exit: () => {
+      assert(
+        instanceAdmin.hasZoeSeatAdmin(zoeSeatAdmin),
+        `Cannot exit seat. Seat has already exited`,
+      );
+      updater.finish(undefined);
+      instanceAdmin.removeZoeSeatAdmin(zoeSeatAdmin);
+
+      /** @type {PaymentPKeywordRecord} */
+      const payout = {};
+      Object.entries(currentAllocation).forEach(([keyword, payoutAmount]) => {
+        const purse = brandToPurse.get(payoutAmount.brand);
+        payout[keyword] = E(purse).withdraw(payoutAmount);
+      });
+      harden(payout);
+      payoutPromiseKit.resolve(payout);
+    },
+    getCurrentAllocation: () => currentAllocation,
+  });
+
+  /** @type {UserSeat} */
+  const userSeat = harden({
+    getCurrentAllocation: async () => zoeSeatAdmin.getCurrentAllocation(),
+    getProposal: async () => proposal,
+    getPayouts: async () => payoutPromiseKit.promise,
+    getPayout: async keyword =>
+      payoutPromiseKit.promise.then(payouts => payouts[keyword]),
+    getOfferResult: async () => offerResultPromiseKit.promise,
+    hasExited: async () => instanceAdmin.hasZoeSeatAdmin(zoeSeatAdmin),
+    exit: async () =>
+      exitObjPromiseKit.promise.then(exitObj => E(exitObj).exit()),
```

Comment:
This code can produce two unpleasant failure behaviors:

If this is an offerless seat, `makeZoeSeatAdminKit` is called with no `exitObj` named argument, resulting in `exitObjPromiseKit.promise` being a promise that never resolved. If one asks an offerless seat to `tryExit`, no response is ever received and no diagnostic is reported anywhere.

If this is a seat whose exit kind is anything other than `'onDemand'`, then `exitObjPromiseKit.promise` will be a promise for undefined. The promise above will eventually reject, but with a completely uninformative diagnostic that looks like an internal error rather than an inappropriate request.

To address the first problem, I suggest that the default exitObj here be one that complains that this is an offerless seat. And the default exitObj in exit.js be one that complains that this exit kind, which it shows, does not allow the user to exit the userSeat on demand.

Reply:
Why force a round trip by
```js
exitObjPromiseKit.promise.then(exitObj => E(exitObj).exit())
```
? Better is simply
```js
E(exitObjPromiseKit.promise).exit()
```

Reply:
I feel like offerless seats should be on demand by default. Hmm, I think that might require some redesign. Until then, I will throw an error message that says that offerless seats cannot be exited.